bazaarvoice/emodb
# Replace guava event bus with a system that propagates exceptions

## Github Issue

None

## What Are We Doing Here?

Emo's `DefaultDataStore` utilizes a Guava `EventBus` to post events to `DefaultDatabus`, which gets events via a subscription. The problem lies in the guarantees of `EventBus`.
```
/* Posts an event to all registered subscribers.
This method will return successfully after the event has been posted to all
subscribers, and regardless of any exceptions thrown by subscribers.
*/
public void post(Object event)
```
This is a huge problem because `EventSus` swallows all exception thrown by its subscribers, and therefore failed writes to the databus will not stop writes to the system of record. This has certainly led to missed events unfortunately.

The solution to this problem is to replace `EventBus` with our own messaging system that propagates failure.

## How to Test and Verify

1. Check out this PR
2. Build and Run and ensure everything is working as expected
3. Modify `writeEvent` in `DefaultDatabus` to throw a `RuntimeException` and fail to write to the databus.
4. Ensure that this prevented the system of record from being updated via CQL.

## Risk

### Level 

`Low`

### Required Testing

`Regression`

### Risk Summary

The only risk here is that default databus doesn't register itself as the event writer and therefore events are discarded. That being said, that risk only lies within a future update to emo that causes such an issue.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
